src/Workspaces/Core/Portable/Storage/AbstractPersistentStorageService.cs

```diff
@@ -258,6 +272,15 @@ public Task<Stream> ReadStreamAsync(Project project, string name, CancellationTo
             public Task<Stream> ReadStreamAsync(Document document, string name, CancellationToken cancellationToken = default)
                 => _storage.Target.ReadStreamAsync(document, name, cancellationToken);
 
+            public Task<Stream> ReadStreamAsync(string name, Checksum checksum = null, CancellationToken cancellationToken = default)
```

Comment:
if we support ReadChecksumAsync. ReadStreamAsync probably doesn't need checksum paramter and and this one can be provided as extension method? that call ReadChecksum and call the other one?

Reply:
>  ReadStreamAsync probably doesn't need checksum paramter

You still need it because of race conditions.  You need to be able to read the stream *conditionally* on the checksum matching.  If you have these as separate calls you could end up with the race where:

1. you call ReadChecksum.  you validate the checksums match.
2. other thread writes to that stream.
3. you call ReadStream.  But the data has changed.

Reply:
so other design is we expose transaction? also, if that is the issue, not provide ReadChecksumAsync at all?

Reply:
I would definitely like to *never* expse transactions.  It will just be far too easy for people to screw things up.  Having an atomic key/value store seems like a much better abstraction for people to have to understand and work with.